File: src/sites/FurAffinity.ts
import BaseSite from './BaseSite';
import ProcessResponse from './ProcessResponse';
import got from 'got';
import { version } from '../../package.json';
import { Message, MessageEmbed } from 'discord.js';
import CacheManager from '../CacheManager';
import { processDescription } from '../Helpers';

class FurAffinity extends BaseSite {
    identifier = 'FurAffinity';

    pattern =
        /https?:\/\/(www\.)?furaffinity\.net\/(?:view|full)\/(?<id>\d+)\/?/gim;

    async process(
        match: RegExpMatchArray,
        /* eslint-disable @typescript-eslint/no-unused-vars */
        source: Message | null
    ): Promise<ProcessResponse | false> {
        const message: ProcessResponse = {
            embeds: [],
            files: [],
        };

        if (!match.groups?.id) {
            return Promise.resolve(false);
        }

        const response = await this.getSubmission(match.groups.id);

        if (!response) {
            return Promise.resolve(false);
        }

        const embed = new MessageEmbed({
            title: response.title,
            url: match[0],
            color: this.color,
            timestamp: response.posted_at,
            description: processDescription(response.description),
            image: {
                url: response.download,
            },
            author: {
                name: response.profile_name,
                url: response.profile,
                iconURL: response.avatar,
            },
            fields: [
                {
                    name: 'Views',
                    value: response.views ?? '0',
                    inline: true,
                },
                {
                    name: 'Favorites',
                    value: response.favorites ?? '0',
                    inline: true,
                },
                {
                    name: 'Comments',
                    value: response.comments ?? '0',
                    inline: true,
                },
            ],
            footer: {
                text: 'FurAffinity',
            },
        });

        message.embeds.push(embed);

        return Promise.resolve(message);
    }

    async getSubmission(id: string): Promise<FAExportSubmission | null> {
        const cacheKey = `furaffinity.submission_${id}`;
        const cacheManager = await CacheManager.getInstance();

        const cachedValue = await cacheManager.remember(cacheKey, async () => {
            const results = await got
                .get(`https://faexport.spangle.org.uk/submission/${id}.json`, {
                    responseType: 'json',
                    headers: {
                        'User-Agent': `SaucyBot/${version}`,
                    },
                })
                .json<FAExportSubmission>();

            return Promise.resolve(JSON.stringify(results));
        });

        if (!cachedValue) {
            return Promise.resolve(null);
        }

        const results = JSON.parse(cachedValue) as FAExportSubmission;

        return Promise.resolve(results);
    }
}

interface FAExportSubmission {
    title: string;
    description: string;
    name: string;
    profile: string;
    profile_name: string;
    avatar: string;
    download: string;
    full: string;
    posted_at: string;
    views: string;
    comments: string;
    favorites: string;
}

export default FurAffinity;

File: SaucyBot/Site/FurAffinity.cs
using System.Text.RegularExpressions;
using Discord;
using Discord.WebSocket;
using SaucyBot.Common;
using SaucyBot.Library.Sites.FurAffinity;
using SaucyBot.Site.Response;

namespace SaucyBot.Site;

public sealed partial class FurAffinity : BaseSite
{
    public override string Identifier => "FurAffinity";

    [GeneratedRegex(@"https?://(www\.)?furaffinity\.net/(?:view|full)/(?<id>\d+)/?", RegexOptions.IgnoreCase | RegexOptions.Multiline)]
    private static partial Regex FurAffinityPattern();

    protected override Regex Pattern => FurAffinityPattern();

    private readonly ILogger<FurAffinity> _logger;
    private readonly IFurAffinityClient _client;

    public FurAffinity(ILogger<FurAffinity> logger, IFurAffinityClient client)
    {
        _logger = logger;
        _client = client;
    }
    
    public override async Task<ProcessResponse?> Process(ProcessRequest request)
    {
        var response = new ProcessResponse();

        var submission = await _client.GetSubmission(
            request.Match.Groups["id"].Value
        );
        
        if (submission is null)
        {
            return null;
        }

        var embed = new EmbedBuilder
        {
            Title = submission.Title,
            Description = Helper.ProcessDescription(FurAffinityBbCodeCleaner.Clean(submission.Description)),
            Color = Color,
            Url = submission.Link,
            ImageUrl = submission.Download,
            Timestamp = DateTimeOffset.Parse(submission.PostedAt),
            Author = new EmbedAuthorBuilder
            {
                Name = submission.ProfileName,
                Url = submission.Profile,
                IconUrl = submission.Avatar,
            },
            Fields = new List<EmbedFieldBuilder>
            {
                new()
                {
                    Name = "Views",
                    Value = submission.Views,
                    IsInline = true,
                },
                new()
                {
                    Name = "Favorties",
                    Value = submission.Favorites,
                    IsInline = true,
                },
                new()
                {
                    Name = "Comments",
                    Value = submission.Comments,
                    IsInline = true,
                }
            },
            Footer = new EmbedFooterBuilder { Text = "FurAffinity" },
        };
        
        response.Embeds.Add(embed.Build());
            
        return response;
    }
}
